mm/2s2h/Rando/ActorBehavior/EnTab.cpp

```diff
@@ -0,0 +1,153 @@
+#include "ActorBehavior.h"
+#include "Rando/MiscBehavior/MiscBehavior.h"
+#include <libultraship/libultraship.h>
+
+extern "C" {
+#include "variables.h"
+#include "overlays/actors/ovl_En_Tab/z_en_tab.h"
+
+s32 func_80BE0D38(Actor* thisx, PlayState* play); // Standard dialogue mscript callback that does bottle check
+}
+
+int32_t EnTab_OverrideBottleCheckCallback(Actor* thisx, PlayState* play) {
+    // Should always return true - as if the player always has an empty bottle!
+    // See func_80BE0D38 in z_en_tab.c
+    return true;
+}
+
+void EnTab_OnOpenShopText(u16* textId, bool* loadFromMessageTable) {
+    RandoSaveCheck milkPurchaseCheck = RANDO_SAVE_CHECKS[RC_MILK_BAR_PURCHASE_MILK];
+    RandoSaveCheck chateauPurchaseCheck = RANDO_SAVE_CHECKS[RC_MILK_BAR_PURCHASE_CHATEAU];
+
+    RandoItemId riMilkPurchase = Rando::ConvertItem(milkPurchaseCheck.randoItemId, RC_MILK_BAR_PURCHASE_MILK);
+    RandoItemId riChateauPurchase = Rando::ConvertItem(chateauPurchaseCheck.randoItemId, RC_MILK_BAR_PURCHASE_CHATEAU);
+
+    auto entry = CustomMessage::LoadVanillaMessageTableEntry(*textId);
+    entry.autoFormat = false;
+
+    entry.msg = "\x02\xC3{item1}\x01 {price1} Rupees\x11"
+                "\x02{item2}\x01 {price2} Rupees\x11"
+                "\x02Nothing";
+
+    std::string itemName1 = "Milk";
+    std::string itemPrice1 = "20";
+    if (!milkPurchaseCheck.cycleObtained) {
+        itemName1 = Rando::StaticData::Items[riMilkPurchase].name;
+        itemPrice1 = std::to_string(milkPurchaseCheck.price);
+    }
+
+    std::string itemName2 = "Chateau Romani";
+    std::string itemPrice2 = "200";
+    if (!chateauPurchaseCheck.cycleObtained) {
+        itemName2 = Rando::StaticData::Items[riChateauPurchase].name;
+        itemPrice2 = std::to_string(chateauPurchaseCheck.price);
+    }
+
+    CustomMessage::Replace(&entry.msg, "{item1}", itemName1);
+    CustomMessage::Replace(&entry.msg, "{item2}", itemName2);
+    CustomMessage::Replace(&entry.msg, "{price1}", itemPrice1);
+    CustomMessage::Replace(&entry.msg, "{price2}", itemPrice2);
+    CustomMessage::EnsureMessageEnd(&entry.msg);
+    CustomMessage::LoadCustomMessageIntoFont(entry);
+    *loadFromMessageTable = false;
+};
+
+void Rando::ActorBehavior::InitEnTabBehavior() {
```

Comment:
You need to change all of your `IS_RANDO`'s here to also check if shops are shuffled.